presto-main/src/main/java/com/facebook/presto/cost/UnnestStatsRule.java

```diff
@@ -0,0 +1,75 @@
+package com.facebook.presto.cost;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.cost.ComposableStatsCalculator.Rule;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.plan.UnnestNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.sql.planner.plan.Patterns.unnest;
+import static java.lang.Double.NaN;
+
+public class UnnestStatsRule
+        implements Rule<UnnestNode>
+{
+    private static final int UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION = 1;
+    @Override
+    public Pattern<UnnestNode> getPattern()
+    {
+        return unnest();
+    }
+
+    @Override
+    public Optional<PlanNodeStatsEstimate> calculate(UnnestNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        // Since we don't have stats for cardinality about the unnest column, we cannot estimate the row count.
+        // However, when the source row count is low, the error would not matter much in query optimization.
+        // Thus we'd still populate the inaccurate numbers just so stats are populated to enable optimization
+        // potential.
+        PlanNodeStatsEstimate.Builder calculatedStats = PlanNodeStatsEstimate.builder();
+        if (sourceStats.getOutputRowCount() <= UPPER_BOUND_ROW_COUNT_FOR_ESTIMATION) {
+            calculatedStats.setOutputRowCount(sourceStats.getOutputRowCount());
```

Comment:
Isn't this way underestimated.  if the source stats have one row, then the unnested stats are going to be at least one row, but will likely be way more.

Reply:
Yes, this will most likely be an underestimation. We can also take the route with a default cardinality. But no matter what the number is, it would be wrong more times than it's correct. So instead of coming up with a guessed cardinality that's more reasonable, I'd rather just set the threshold where this applies to a very low number (1 here). If we only estimate row count when the input has only 1 row, it doesn't matter how much we mess up the cardinality. Even if this would actually produce 1 million rows.

I wish we could have a separate way to describe that the total data size will not change dramatically each though the row count would change and the average row size would change. That seems to be a much bigger change to do and I don't have the background to know whether that's generally useful or not. At the same time, I think the implementation here is reasonable in the sense that:
1) when the input has only 1 row, the estimated result would be similar or moderately higher (in terms of map or row type) that the real stats.
2) the logic for estimating column stats is quite conservative and reasonable regardless (note this is difference from the estimation of overall costs).

Reply:
> Even if this would actually produce 1 million rows.
> I wish we could have a separate way to describe that the total data size will not change dramatically

This is only true for a single column use case. Frankly i would like to keep this rule enabled only for that case explicitly.